### PR TITLE
Support HTML Figures in Chromium

### LINKED ISSUE #16137: HTML figures not properly supported in Chromium
### Steps to reproduce:

Load a page containing an HTML or ARIA figure, [such as this demo page](https://saschacowley.com/demos/figure.html), in a Chromium-based browser. Read/navigate the document in browse mode, and observe the output.

### Actual behavior:

Example output with the above-linked page in Edge:

> down arrow  
> elementum eros tempus.  
> down arrow  
> has details  graphic    Marble sculpture of Cleopatra's head  
> down arrow  
> caption    An ancient Roman bust of Ptolemaic ruler Cleopatra VII of Egypt dated to the mid-1st century BCE.   
> down arrow  
> caption    Discovered in a villa along the Via Appia. Now located in the Altes Museum, Berlin.  
> down arrow  
> Etiam fringilla et arcu nec finibus. Phasellus viverra 

### Expected behavior:

Figures should be reported more like they are in Gecko-based browsers.

Example output with the above-linked page in Firefox:

> down arrow  
> elementum eros tempus.  
> down arrow  
> figure    graphic    Marble sculpture of Cleopatra's head  
> down arrow  
> caption    An ancient Roman bust of Ptolemaic ruler Cleopatra VII of Egypt dated to the mid-1st century BCE.   
> down arrow  
> caption    Discovered in a villa along the Via Appia. Now located in the Altes Museum, Berlin.  
> down arrow  
> out of figure  Etiam fringilla et arcu nec finibus. Phasellus viverra massa nec congue interdum. Vivamus placerat 

### NVDA logs, crash dumps and other attachments:

- [Example page](https://saschacowley.com/demos/figure.html)

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

2024.1beta7 (2024.1.0.30959)

#### Windows version:

23H2 (OS Build 22631.3007)

#### Name and version of other software in use when reproducing the issue:

* Microsoft Edge version 121.0.2277.98 (Official build) (64-bit)
* Google Chrome version 121.0.6167.161 (Official Build) (64-bit)

#### Other information about your system:

Nvda Settings -> Advanced -> Use UIA with Microsoft Edge and other 
Chromium based browsers when available: Default (Only when necessary)

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

Issue exists in 2023.3.3

#### If NVDA add-ons are disabled, is your problem still occurring?

Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

Yes

### PR DESCRIPTION
**Pull request by SaschaCowley**
### Link to issue number:

Fixes #16137 

### Summary of the issue:

Figures are currently not properly announced by NVDA in browse or focus modes in Chromium.

### Description of user facing changes

Figures are now reported in Chrome.

### Description of development approach

Since Chromium does not report a role of `figure` on `<figure>` elements unless the role has been added, add tests for a `tag` of `figure` on Chrome objects. Change figure searchable attributes to include the figure tag, since this is necessary to allow navigation by HTML figures in Chrome.

### Testing strategy:

Manual tests with a [test webpage](https://saschacowley.com/demos/figure.html).

### Known issues with pull request:

* Quick navigation by figure will land on `<figure>` elements whose `role` has been set to something other than `figure`.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: source/NVDAObjects/IAccessible/chromium.py [hunk 1]
```diff
@@ -65,6 +65,12 @@ def _normalizeControlField(self, attrs):
 		if attrs['role'] == controlTypes.Role.TOGGLEBUTTON and controlTypes.State.CHECKABLE in attrs['states']:
 			# In Chromium, the checkable state is exposed erroneously on toggle buttons.
 			attrs['states'].discard(controlTypes.State.CHECKABLE)
+
+		if (
+			attrs['role'] == controlTypes.Role.GROUPING
+			and attrs.get('IAccessible2::attribute_tag', "").lower() == "figure"):
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/chromium.py:71]: More standard lint

```suggestion
		if (
			attrs["role"] == controlTypes.Role.GROUPING
			and attrs.get("IAccessible2::attribute_tag", "").lower() == "figure"
		):
```

File: source/NVDAObjects/IAccessible/chromium.py [hunk 2]
```diff
@@ -154,6 +160,12 @@ def _get_states(self):
 		return states
 
 
+class Figure(ia2Web.Ia2Web):
+	def _get_role(self):
+		return controlTypes.Role.FIGURE
+		# return super()._get_role()
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/chromium.py:166]:
```suggestion
```

File: source/NVDAObjects/IAccessible/chromium.py [hunk 3]
```diff
@@ -154,6 +160,12 @@ def _get_states(self):
 		return states
 
 
+class Figure(ia2Web.Ia2Web):
+	def _get_role(self):
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/chromium.py:164]:
```suggestion
	def _get_role(self) -> controlTypes.Role:
```

File: source/NVDAObjects/IAccessible/chromium.py [hunk 4]
```diff
@@ -65,6 +65,12 @@ def _normalizeControlField(self, attrs):
 		if attrs['role'] == controlTypes.Role.TOGGLEBUTTON and controlTypes.State.CHECKABLE in attrs['states']:
 			# In Chromium, the checkable state is exposed erroneously on toggle buttons.
 			attrs['states'].discard(controlTypes.State.CHECKABLE)
+
+		if (
+			attrs['role'] == controlTypes.Role.GROUPING
+			and attrs.get('IAccessible2::attribute_tag', "").lower() == "figure"):
+			# Chromium doesn't expose the `<figure>` element as a figure.
+			attrs['role'] = controlTypes.Role.FIGURE
```
Comment (seanbudd) [inline source/NVDAObjects/IAccessible/chromium.py:73]:
```suggestion
			attrs["role"] = controlTypes.Role.FIGURE
```

File: source/virtualBuffers/gecko_ia2.py [hunk 5]
```diff
@@ -552,7 +552,9 @@ def _searchableAttribsForNodeType(self,nodeType):
 		elif nodeType == "figure":
 			attrs = [
 				{"Iaccessible::role": [oleacc.ROLE_SYSTEM_GROUPING]},
-				{"IAccessible2::attribute_xml-roles": [VBufStorage_findMatch_word("figure")]}
+				{"IAccessible2::attribute_xml-roles": [VBufStorage_findMatch_word("figure")]},
+				# Needed so that navigation by figure works for HTML figures in Chromium
+				{"IAccessible2::attribute_tag": self._searchableTagValues(["figure"])}
```
Comment (seanbudd) [inline source/virtualBuffers/gecko_ia2.py:557]: Terminating lists with commas minimize diffs like this where you add an item

```suggestion
				{"IAccessible2::attribute_tag": self._searchableTagValues(["figure"])},
```

**Review by seanbudd — APPROVED** [review]
Thanks @SaschaCowley. As the testing field is unticked, can you confirm you've tested this?

**Comment by SaschaCowley** [comment]
> Thanks @SaschaCowley. As the testing field is unticked, can you confirm you've tested this?

This has been manually tested. The initial PR comment has been updated to reflect this.

**Pull request merged**